### PR TITLE
Fix FileUplink packet sequence repeat and CRC #1378

### PR DESCRIPTION
**Pull request by DJKessler**
| | |
|:---|:---|
|**_Related Issue(s)_**| #1378 |
|**_Has Unit Tests (y/n)_**| y |
|**_Documentation Included (y/n)_**| n |

---
## Change Description

Added duplicate packet detection based on sequence number

- Added duplicated packet event/warning
- Added method to check for duplicated packet
- Added last packet write status `FileUplink` member variable
- Duplicate packets are now detected and skipped
- Added duplicate file packet unit test

### CODE REVIEW

**Review by LeStarch — APPROVED** [review]
This is what I see:  if the last packet was successful *and* this packet is a duplicate, then the rest of the function (writing, CRC) is skipped.  To my estimate this completes the work!

**Comment by LeStarch** [comment]
@timcanham would you review too, given you were more involved with this work?

File: Svc/FileUplink/FileUplink.cpp [hunk 1]
```diff
@@ -174,6 +185,18 @@ namespace Svc {
     this->m_lastSequenceIndex = sequenceIndex;
   }
 
+  bool FileUplink ::
+    checkDuplicatedPacket(const U32 sequenceIndex)
+  {
+    // check for duplicate packet
+    if (sequenceIndex == this->m_lastSequenceIndex) {
+      this->m_warnings.packetDuplicate(sequenceIndex);
```
Comment (timcanham) [inline Svc/FileUplink/FileUplink.cpp:193]: How do you avoid a duplicate on the first packet if the first packet sequence number is 0?

Comment (DJKessler) [reply Svc/FileUplink/FileUplink.cpp:193]: Fair point. I suppose `m_lastSequenceIndex` should be initialized to some non-zero value that also won't be a valid packet index (eg. `U32MAX`).

Comment (timcanham) [reply Svc/FileUplink/FileUplink.cpp:193]: I think it would actually drop the first packet, since the last status was OK.

Comment (DJKessler) [reply Svc/FileUplink/FileUplink.cpp:193]: You're right; `m_lastPacketWriteStatus` should be initialized to something other than `Os::File::OP_OK`.

Comment (timcanham) [reply Svc/FileUplink/FileUplink.cpp:193]: Whatever solution you come up with, you should try running it against the ground system and uplink a file nominally to make sure something doesn't get through.

**Review by timcanham — APPROVED** [review]
Looks good!

**Comment by LeStarch** [comment]
Thanks @DJKessler!

**Pull request merged**